### PR TITLE
[WIP] Fix tray download location in crc home folder

### PR DESCRIPTION
**Pull request by anjannath**
**Fixes:** Issue #1275 

## Testing

`crc setup --enable-experimental-features` should start the tray and daemon services successfully

### CODE REVIEW

File: pkg/crc/cache/cache_tray_windows.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/code-ready/crc/pkg/extract"
+	crcos "github.com/code-ready/crc/pkg/os"
+	"github.com/pkg/errors"
+)
+
+type TrayCache struct {
+	Cache
+}
+
+func (tc *TrayCache) CacheBinary() error {
```
Comment (anjannath) [inline pkg/crc/cache/cache_tray_windows.go:18]: The original `CacheBinary()` function only extracts a single binary that matches the filter, for the tray we want to extract the complete directory

Comment (cfergeau) [reply pkg/crc/cache/cache_tray_windows.go:18]: I don't think you need to duplicate the whole function when there's only a slight difference:
`extractedFiles, err := extract.Uncompress(assetTmpFile, tmpDir)` VS `extractedFiles, err := extract.UncompressWithFilter(assetTmpFile, tmpDir, ...)` You could have an internal helper with more args (a file filter) which would be used by both `CacheBinary` and `TrayCache.CacheBinary`. I don't think you need `TrayCache` at all, it could probably be called `TrayCache.CacheBinary` could probably be `Cache.CacheDir` or something like that?

File: pkg/crc/cache/cache.go [hunk 2]
```diff
@@ -20,8 +20,8 @@ type Cache struct {
 	destDir    string
 }
 
-func New(binarName string, archiveURL string, destDir string) *Cache {
-	return &Cache{binaryName: binarName, archiveURL: archiveURL, destDir: destDir}
+func New(binaryName string, archiveURL string, destDir string) *Cache {
+	return &Cache{binaryName: binaryName, archiveURL: archiveURL, destDir: destDir}
 }
```
Comment (cfergeau) [inline pkg/crc/cache/cache.go:25]: nit: This typo fix is totally unrelated to these changes, only mentioning it since it took me a while to notice `binarName` vs `binaryName`

File: pkg/crc/preflight/preflight_checks_tray_windows.go [hunk 3]
```diff
@@ -148,6 +113,12 @@ func checkTrayBinaryVersion() error {
 }
 
 func fixTrayBinaryVersion() error {
+	// Kill older running tray
+	if err := checkTrayRunning(); err == nil {
+		if err := stopTray(); err != nil {
+			logging.Debugf("Failed to kill running tray: %v", err)
+		}
+	}
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_checks_tray_windows.go:121]: This code used to be in `fixTrayBinaryExists`, not clear why it moved.

File: pkg/crc/cache/cache_tray_windows.go (hunk 1 above)
Comment (anjannath) [reply pkg/crc/cache/cache_tray_windows.go:18]: I avoided to add a `cache.CacheDir` because i wasn't sure if calling it on the other objects like `podmanCache` would result in a side effect..

File: pkg/crc/cache/cache.go (hunk 2 above)
Comment (anjannath) [reply pkg/crc/cache/cache.go:25]: will put this is in a separate commit

File: pkg/crc/preflight/preflight_checks_tray_windows.go (hunk 3 above)
Comment (anjannath) [reply pkg/crc/preflight/preflight_checks_tray_windows.go:121]: Yes, but it should have been here all along, we can only expect a running tray when there is one available, and only when it is available and the versions don't match we need to update the version in which case we'll have to stop the running tray, in normal case when `fixTrayBinaryExists` is called, it'll most probably not have a running tray, as that binary doesn't exist yet

File: pkg/crc/cache/cache_tray_windows.go (hunk 1 above)
Comment (anjannath) [reply pkg/crc/cache/cache_tray_windows.go:18]: @cfergeau Is this what you were suggesting..
```diff
diff --git a/pkg/crc/cache/cache.go b/pkg/crc/cache/cache.go
index 72a5bab..28cf083 100644
--- a/pkg/crc/cache/cache.go
+++ b/pkg/crc/cache/cache.go
@@ -20,8 +20,8 @@ type Cache struct {
 	destDir    string
 }
 
-func New(binarName string, archiveURL string, destDir string) *Cache {
-	return &Cache{binaryName: binarName, archiveURL: archiveURL, destDir: destDir}
+func New(binaryName string, archiveURL string, destDir string) *Cache {
+	return &Cache{binaryName: binaryName, archiveURL: archiveURL, destDir: destDir}
 }
 
 func NewOcCache(destDir string) *Cache {
@@ -58,7 +58,20 @@ func (c *Cache) CacheBinary() error {
 	if c.IsCached() {
 		return nil
 	}
+	return c.cache(func(filename string) bool {
+		return filepath.Base(filename) == c.binaryName
+	})
+}
 
+// CacheDir downloads and caches the requested asset's dir into crc home
+func (c *Cache) CacheDir() error {
+	if c.IsCached() {
+		return nil
+	}
+	return c.cache(nil)
+}
+
+func (c *Cache) cache(fileFilter func(string) bool) error {
 	// Create tmp dir to download the requested tarball
 	tmpDir, err := ioutil.TempDir("", "crc")
 	if err != nil {
@@ -70,9 +83,13 @@ func (c *Cache) CacheBinary() error {
 		return err
 	}
 
-	// Extract the tarball and put it the cache directory.
-	extractedFiles, err := extract.UncompressWithFilter(assetTmpFile, tmpDir,
-		func(filename string) bool { return filepath.Base(filename) == c.binaryName })
+	// Extract the tarball and put it in the CrcBin directory.
+	extractedFiles, err := extract.UncompressWithFilter(
+		assetTmpFile,
+		tmpDir,
+		fileFilter,
+	)
+
 	if err != nil {
 		return errors.Wrapf(err, "Cannot uncompress '%s'", assetTmpFile)
 	}
diff --git a/pkg/crc/cache/cache_tray_windows.go b/pkg/crc/cache/cache_tray_windows.go
new file mode 100644
index 0000000..17010db
--- /dev/null
+++ b/pkg/crc/cache/cache_tray_windows.go
@@ -0,0 +1,7 @@
+package cache
+
+import "github.com/code-ready/crc/pkg/crc/constants"
+
+func NewWindowsTrayCache(destDir string) *Cache {
+	return New(constants.TrayBinaryName, constants.GetCRCWindowsTrayDownloadURL(), destDir)
+}

```

File: pkg/crc/preflight/preflight_checks_tray_windows.go (hunk 3 above)
Comment (cfergeau) [reply pkg/crc/preflight/preflight_checks_tray_windows.go:121]: can you move this to its own commit? The change to use `Cache` should be purely code moving around, with no behaviour change, the explanation you just gave deserves to be in a git commit log :)

File: pkg/crc/cache/cache_tray_windows.go (hunk 1 above)
Comment (cfergeau) [reply pkg/crc/cache/cache_tray_windows.go:18]: > @cfergeau Is this what you were suggesting..

If `UncompressWithFilter` knows how to deal with a `nil` filter, yes, looks like it! Maybe move the `c.IsCached()` to the helper too?

> The original `CacheBinary()` function only extracts a single binary that matches the filter, for the tray we want to extract the complete directory

Did not occur to me initially, but this can also be added to the commit log, so that future you (or other readers) can remember why this was done this way. What is in that directory though in addition to the tray binary?

> I avoided to add a `cache.CacheDir` because i wasn't sure if calling it on the other objects like `podmanCache` would result in a side effect..

Hmm if someone tries to use this on an inappropriate instance, I would say it's their problem/their bug. We want to avoid non-generic code in this package (OcCache/PodmanCache/... should not have been added there, this was requested during the review but slipped through), so adding a non-generic TrayCache class is going the wrong way. So let's try to add the helper you need to the base class, we can refine the API later if it's not good enough.

Comment (anjannath) [reply pkg/crc/cache/cache_tray_windows.go:18]: > What is in that directory though in addition to the tray binary?

In case of windows there's a dll of a json parser library, that am using and an app config file, in case of macOS the whole app is actually a directory, but macOS would recognize it as an app bundle, it also contains some app manifests, static resources like the icons.

Comment (anjannath) [reply pkg/crc/cache/cache_tray_windows.go:18]: @cfergeau one thing i missed in that diff, we'd also need something like `cache.EnsureDirIsCached`

**Comment by anjannath** [comment]
Closing as this might need more work after #1282
